src/charm.py

```diff
@@ -6,20 +6,355 @@
 
 """A Juju charm for configuring Charmed Ory Oathkeeper with downstream applications."""
 
+import json
 import logging
+import textwrap
+from dataclasses import dataclass
+from itertools import starmap
+from typing import Any, Iterable, Optional
+from urllib.parse import urlparse
 
-from ops.charm import CharmBase
+import jinja2
+from charms.traefik_k8s.v1.ingress_per_unit import (
+    IngressDataReadyEvent,
+    IngressPerUnitProvider,
+    RequirerData,
+)
+from ops.charm import CharmBase, ConfigChangedEvent, EventBase
+from ops.framework import StoredState
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, Relation
+
+from types_ import TraefikConfig, UnitConfig
 
 logger = logging.getLogger(__name__)
 
 
+class InvalidConfigError(Exception):
+    """Internal exception that is raised if the charm config is not valid."""
+
+
+class RuleDerivationError(RuntimeError):
+    """Raised when a rule cannot be derived from other config parameters.
+
+    Solution: provide the rule manually, or fix what's broken.
+    """
+
+    def __init__(self, url: str, *args):
+        msg = f"Unable to derive Rule from {url!r}; ensure that the url is valid."
+        super().__init__(msg, *args)
+
+
+class TemplateKeyError(RuntimeError):
+    """Raised when a template contains a key which we cannot provide.
+
+    Solution: fix the template to only include the variables:
+        - `juju_model`
+        - `juju_application`
+        - `juju_unit`
+    """
+
+    def __init__(self, template: str, key: str, *args):
+        msg = textwrap.dedent(
+            f"""Unable to render the template {template!r}: {key!r} unknown.
+                - `juju_model`
+                - `juju_application`
+                - `juju_unit`"""
+        )
+        super().__init__(msg, *args)
+
+
+@dataclass
+class RouteConfig:
+    """Route configuration."""
+
+    root_url: str
+    rule: str
+    id_: str
+
+
+@dataclass
+class _RouteConfig:
+    root_url: str
+    rule: Optional[str] = None
+
+    @property
+    def is_valid(self) -> bool:
+        def _check_var(obj: str, name: str) -> bool:
+            error = None
+            if not _is_not_empty(obj):
+                error = (
+                    f"`{name}` not configured; run `juju config oathkeeper-configurator"
+                    f"{name}=<{name.upper()}>"
+                )
+
+            elif obj != (stripped := obj.strip()):
```

Comment:
Maybe `str.startswith` and `str.endswith`?

Another thing is that:
Is this rule very important to us in the biz logic? If not, probably it would be better if we could skip verifying this by normalizing the attribute based on some rules (e.g. do not start or end with whitespaces) for users. We can normalize it when we set the attribute using the descriptor.